<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get genomic ranges based on specified features and flanking regions. — getFeature • seqEasyFunctions</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get genomic ranges based on specified features and flanking regions. — getFeature"><meta name="description" content="This function takes a GRanges object and extracts or creates new genomic
ranges based on specified start and end features (TSS, TES, or Exon) and
their flanking regions. It allows for specifying exon numbers and boundaries
for precise range definition."><meta property="og:description" content="This function takes a GRanges object and extracts or creates new genomic
ranges based on specified start and end features (TSS, TES, or Exon) and
their flanking regions. It allows for specifying exon numbers and boundaries
for precise range definition."><meta property="og:image" content="/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">seqEasyFunctions</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/zackswales/seqEasyFunctions">GitHub</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Get genomic ranges based on specified features and flanking regions.</h1>

      <div class="d-none name"><code>getFeature.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function takes a GRanges object and extracts or creates new genomic
ranges based on specified start and end features (TSS, TES, or Exon) and
their flanking regions. It allows for specifying exon numbers and boundaries
for precise range definition.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getFeature</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  start_feature <span class="op">=</span> <span class="st">"TSS"</span>,</span>
<span>  start_flank <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  start_exon <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  start_exon_boundary <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  start_direction <span class="op">=</span> <span class="st">"up"</span>,</span>
<span>  end_feature <span class="op">=</span> <span class="st">"TES"</span>,</span>
<span>  end_flank <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  end_direction <span class="op">=</span> <span class="st">"down"</span>,</span>
<span>  end_exon <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  end_exon_boundary <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"all"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A <code>GRanges</code> object containing genomic features (e.g., genes
with exons). This object must have metadata columns including 'blockStarts',
'blockSizes', and 'blockCount' if 'Exon' is specified as a feature.</p></dd>


<dt id="arg-start-feature">start_feature<a class="anchor" aria-label="anchor" href="#arg-start-feature"></a></dt>
<dd><p>A character string specifying the starting feature.
Must be one of "TSS" (Transcription Start Site), "TES" (Transcription
End Site), or "Exon". Default is "TSS".</p></dd>


<dt id="arg-start-flank">start_flank<a class="anchor" aria-label="anchor" href="#arg-start-flank"></a></dt>
<dd><p>An integer indicating the number of bases to flank the
<code>start_feature</code>. Default is 0.</p></dd>


<dt id="arg-start-exon">start_exon<a class="anchor" aria-label="anchor" href="#arg-start-exon"></a></dt>
<dd><p>An integer specifying the exon number to use if
<code>start_feature</code> is "Exon".</p></dd>


<dt id="arg-start-exon-boundary">start_exon_boundary<a class="anchor" aria-label="anchor" href="#arg-start-exon-boundary"></a></dt>
<dd><p>A character string specifying the exon boundary to
use if <code>start_feature</code> is "Exon". Must be either "5prime" or
"3prime".</p></dd>


<dt id="arg-start-direction">start_direction<a class="anchor" aria-label="anchor" href="#arg-start-direction"></a></dt>
<dd><p>A character string indicating the direction of the
flank relative to the <code>start_feature</code>. Must be either "up"
(upstream) or "down" (downstream). Default is "up".</p></dd>


<dt id="arg-end-feature">end_feature<a class="anchor" aria-label="anchor" href="#arg-end-feature"></a></dt>
<dd><p>A character string specifying the ending feature.
Must be one of "TSS", "TES", or "Exon". Default is "TES".</p></dd>


<dt id="arg-end-flank">end_flank<a class="anchor" aria-label="anchor" href="#arg-end-flank"></a></dt>
<dd><p>An integer indicating the number of bases to flank the
<code>end_feature</code>. Default is 0.</p></dd>


<dt id="arg-end-direction">end_direction<a class="anchor" aria-label="anchor" href="#arg-end-direction"></a></dt>
<dd><p>A character string indicating the direction of the
flank relative to the <code>end_feature</code>. Must be either "up"
(upstream) or "down" (downstream). Default is "down".</p></dd>


<dt id="arg-end-exon">end_exon<a class="anchor" aria-label="anchor" href="#arg-end-exon"></a></dt>
<dd><p>An integer specifying the exon number to use if
<code>end_feature</code> is "Exon".</p></dd>


<dt id="arg-end-exon-boundary">end_exon_boundary<a class="anchor" aria-label="anchor" href="#arg-end-exon-boundary"></a></dt>
<dd><p>A character string specifying the exon boundary to
use if <code>end_feature</code> is "Exon". Must be either "5prime" or
"3prime".</p></dd>


<dt id="arg-return">return<a class="anchor" aria-label="anchor" href="#arg-return"></a></dt>
<dd><p>A character string specifying which ranges to return. Must be
one of "all" (both plus and minus strand ranges), "plus" (only plus
strand ranges), or "minus" (only minus strand ranges). Default is "all".</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>GRanges</code> object containing the extracted or created genomic ranges.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Create a dummy GRanges object</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     tapply, union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:utils’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     findMatches</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  seqnames <span class="op">=</span> <span class="st">"chr1"</span>,</span></span>
<span class="r-in"><span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">2000</span><span class="op">)</span>, end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">3000</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  blockStarts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0,200,500"</span>, <span class="st">"0,300"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  blockSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"100,300,500"</span>, <span class="st">"200,500"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  blockCount <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get TSS regions with 100bp upstream flank</span></span></span>
<span class="r-in"><span><span class="va">tss_regions</span> <span class="op">&lt;-</span> <span class="fu">getFeature</span><span class="op">(</span><span class="va">genes</span>, start_feature <span class="op">=</span> <span class="st">"TSS"</span>, start_flank <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tss_regions</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand | blockStarts  blockSizes blockCount</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1 2000-3100      - |       0,300     200,500          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1    0-1000      + |   0,200,500 100,300,500          3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get regions from TSS to TES</span></span></span>
<span class="r-in"><span><span class="va">gene_bodies</span> <span class="op">&lt;-</span> <span class="fu">getFeature</span><span class="op">(</span><span class="va">genes</span>, start_feature <span class="op">=</span> <span class="st">"TSS"</span>, end_feature <span class="op">=</span> <span class="st">"TES"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene_bodies</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand | blockStarts  blockSizes blockCount</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1 2000-3000      - |       0,300     200,500          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1  100-1000      + |   0,200,500 100,300,500          3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get regions around the 2nd exon start</span></span></span>
<span class="r-in"><span><span class="va">exon2_start</span> <span class="op">&lt;-</span> <span class="fu">getFeature</span><span class="op">(</span><span class="va">genes</span>, start_feature <span class="op">=</span> <span class="st">"Exon"</span>, start_exon <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>                           start_exon_boundary <span class="op">=</span> <span class="st">"5prime"</span>, start_flank <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">exon2_start</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand | blockStarts  blockSizes blockCount</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1 2000-2250      - |       0,300     200,500          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1  250-1000      + |   0,200,500 100,300,500          3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get regions spanning the 1st to the 2nd exon</span></span></span>
<span class="r-in"><span><span class="va">exon1_to_2</span> <span class="op">&lt;-</span> <span class="fu">getFeature</span><span class="op">(</span><span class="va">genes</span>, start_feature <span class="op">=</span> <span class="st">"Exon"</span>, start_exon <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                          start_exon_boundary <span class="op">=</span> <span class="st">"5prime"</span>,</span></span>
<span class="r-in"><span>                          end_feature <span class="op">=</span> <span class="st">"Exon"</span>, end_exon <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>                          end_exon_boundary <span class="op">=</span> <span class="st">"3prime"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">exon1_to_2</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand | blockStarts  blockSizes blockCount</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1 2000-2800      - |       0,300     200,500          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1   100-600      + |   0,200,500 100,300,500          3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Zack Swales.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

