<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Get genomic ranges based on specified features and flanking regions. — getFeature • seqEasyFunctions</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Get genomic ranges based on specified features and flanking regions. — getFeature"><meta property="og:description" content="This function takes a GRanges object and extracts or creates new genomic
ranges based on specified start and end features (TSS, TES, or Exon) and
their flanking regions. It allows for specifying exon numbers and boundaries
for precise range definition."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqEasyFunctions</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Get genomic ranges based on specified features and flanking regions.</h1>

    <div class="hidden name"><code>getFeature.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes a GRanges object and extracts or creates new genomic
ranges based on specified start and end features (TSS, TES, or Exon) and
their flanking regions. It allows for specifying exon numbers and boundaries
for precise range definition.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">getFeature</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  start_feature <span class="op">=</span> <span class="st">"TSS"</span>,</span>
<span>  start_flank <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  start_exon <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  start_exon_boundary <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  start_direction <span class="op">=</span> <span class="st">"up"</span>,</span>
<span>  end_feature <span class="op">=</span> <span class="st">"TES"</span>,</span>
<span>  end_flank <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  end_direction <span class="op">=</span> <span class="st">"down"</span>,</span>
<span>  end_exon <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  end_exon_boundary <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  return <span class="op">=</span> <span class="st">"all"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A <code>GRanges</code> object containing genomic features (e.g., genes
with exons). This object must have metadata columns including 'blockStarts',
'blockSizes', and 'blockCount' if 'Exon' is specified as a feature.</p></dd>


<dt id="arg-start-feature">start_feature<a class="anchor" aria-label="anchor" href="#arg-start-feature"></a></dt>
<dd><p>A character string specifying the starting feature.
Must be one of "TSS" (Transcription Start Site), "TES" (Transcription
End Site), or "Exon". Default is "TSS".</p></dd>


<dt id="arg-start-flank">start_flank<a class="anchor" aria-label="anchor" href="#arg-start-flank"></a></dt>
<dd><p>An integer indicating the number of bases to flank the
<code>start_feature</code>. Default is 0.</p></dd>


<dt id="arg-start-exon">start_exon<a class="anchor" aria-label="anchor" href="#arg-start-exon"></a></dt>
<dd><p>An integer specifying the exon number to use if
<code>start_feature</code> is "Exon".</p></dd>


<dt id="arg-start-exon-boundary">start_exon_boundary<a class="anchor" aria-label="anchor" href="#arg-start-exon-boundary"></a></dt>
<dd><p>A character string specifying the exon boundary to
use if <code>start_feature</code> is "Exon". Must be either "5prime" or
"3prime".</p></dd>


<dt id="arg-start-direction">start_direction<a class="anchor" aria-label="anchor" href="#arg-start-direction"></a></dt>
<dd><p>A character string indicating the direction of the
flank relative to the <code>start_feature</code>. Must be either "up"
(upstream) or "down" (downstream). Default is "up".</p></dd>


<dt id="arg-end-feature">end_feature<a class="anchor" aria-label="anchor" href="#arg-end-feature"></a></dt>
<dd><p>A character string specifying the ending feature.
Must be one of "TSS", "TES", or "Exon". Default is "TES".</p></dd>


<dt id="arg-end-flank">end_flank<a class="anchor" aria-label="anchor" href="#arg-end-flank"></a></dt>
<dd><p>An integer indicating the number of bases to flank the
<code>end_feature</code>. Default is 0.</p></dd>


<dt id="arg-end-direction">end_direction<a class="anchor" aria-label="anchor" href="#arg-end-direction"></a></dt>
<dd><p>A character string indicating the direction of the
flank relative to the <code>end_feature</code>. Must be either "up"
(upstream) or "down" (downstream). Default is "down".</p></dd>


<dt id="arg-end-exon">end_exon<a class="anchor" aria-label="anchor" href="#arg-end-exon"></a></dt>
<dd><p>An integer specifying the exon number to use if
<code>end_feature</code> is "Exon".</p></dd>


<dt id="arg-end-exon-boundary">end_exon_boundary<a class="anchor" aria-label="anchor" href="#arg-end-exon-boundary"></a></dt>
<dd><p>A character string specifying the exon boundary to
use if <code>end_feature</code> is "Exon". Must be either "5prime" or
"3prime".</p></dd>


<dt id="arg-return">return<a class="anchor" aria-label="anchor" href="#arg-return"></a></dt>
<dd><p>A character string specifying which ranges to return. Must be
one of "all" (both plus and minus strand ranges), "plus" (only plus
strand ranges), or "minus" (only minus strand ranges). Default is "all".</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>GRanges</code> object containing the extracted or created genomic ranges.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Create a dummy GRanges object</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     tapply, union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:utils’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     findMatches</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: IRanges</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: GenomeInfoDb</span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  seqnames <span class="op">=</span> <span class="st">"chr1"</span>,</span></span>
<span class="r-in"><span>  ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">2000</span><span class="op">)</span>, end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">3000</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  blockStarts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0,200,500"</span>, <span class="st">"0,300"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  blockSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"100,300,500"</span>, <span class="st">"200,500"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  blockCount <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get TSS regions with 100bp upstream flank</span></span></span>
<span class="r-in"><span><span class="va">tss_regions</span> <span class="op">&lt;-</span> <span class="fu">getFeature</span><span class="op">(</span><span class="va">genes</span>, start_feature <span class="op">=</span> <span class="st">"TSS"</span>, start_flank <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tss_regions</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand | blockStarts  blockSizes blockCount</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1 2000-3100      - |       0,300     200,500          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1    0-1000      + |   0,200,500 100,300,500          3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get regions from TSS to TES</span></span></span>
<span class="r-in"><span><span class="va">gene_bodies</span> <span class="op">&lt;-</span> <span class="fu">getFeature</span><span class="op">(</span><span class="va">genes</span>, start_feature <span class="op">=</span> <span class="st">"TSS"</span>, end_feature <span class="op">=</span> <span class="st">"TES"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene_bodies</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand | blockStarts  blockSizes blockCount</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1 2000-3000      - |       0,300     200,500          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1  100-1000      + |   0,200,500 100,300,500          3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get regions around the 2nd exon start</span></span></span>
<span class="r-in"><span><span class="va">exon2_start</span> <span class="op">&lt;-</span> <span class="fu">getFeature</span><span class="op">(</span><span class="va">genes</span>, start_feature <span class="op">=</span> <span class="st">"Exon"</span>, start_exon <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>                           start_exon_boundary <span class="op">=</span> <span class="st">"5prime"</span>, start_flank <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">exon2_start</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand | blockStarts  blockSizes blockCount</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1 2000-2250      - |       0,300     200,500          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1  250-1000      + |   0,200,500 100,300,500          3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get regions spanning the 1st to the 2nd exon</span></span></span>
<span class="r-in"><span><span class="va">exon1_to_2</span> <span class="op">&lt;-</span> <span class="fu">getFeature</span><span class="op">(</span><span class="va">genes</span>, start_feature <span class="op">=</span> <span class="st">"Exon"</span>, start_exon <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                          start_exon_boundary <span class="op">=</span> <span class="st">"5prime"</span>,</span></span>
<span class="r-in"><span>                          end_feature <span class="op">=</span> <span class="st">"Exon"</span>, end_exon <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>                          end_exon_boundary <span class="op">=</span> <span class="st">"3prime"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">exon1_to_2</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 2 ranges and 3 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       seqnames    ranges strand | blockStarts  blockSizes blockCount</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1]     chr1 2000-2800      - |       0,300     200,500          2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2]     chr1   100-600      + |   0,200,500 100,300,500          3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Zack Swales.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

