<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate a list of normalized matrices from BigWig files over genomic regions. — matList • seqEasyFunctions</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a list of normalized matrices from BigWig files over genomic regions. — matList"><meta property="og:description" content="This function takes forward and reverse strand BigWig files and a list of genomic region sets,
then returns a list of normalized matrices representing signal intensities over those regions.
It supports flexible control over binning, strand handling, normalization mode, smoothing,
and how the matrix represents target regions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">seqEasyFunctions</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate a list of normalized matrices from BigWig files over genomic regions.</h1>

    <div class="hidden name"><code>matList.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes forward and reverse strand BigWig files and a list of genomic region sets,
then returns a list of normalized matrices representing signal intensities over those regions.
It supports flexible control over binning, strand handling, normalization mode, smoothing,
and how the matrix represents target regions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">matList</span><span class="op">(</span></span>
<span>  <span class="va">bwf</span>,</span>
<span>  <span class="va">bwr</span>,</span>
<span>  <span class="va">names</span>,</span>
<span>  <span class="va">grl</span>,</span>
<span>  wins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Gene <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"coverage"</span>,</span>
<span>  output <span class="op">=</span> <span class="st">"norm.matrix"</span>,</span>
<span>  strand <span class="op">=</span> <span class="st">"rev"</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  extend <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  <span class="va">w</span>,</span>
<span>  include_target <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  target_ratio <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  <span class="va">keep</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-bwf">bwf<a class="anchor" aria-label="anchor" href="#arg-bwf"></a></dt>
<dd><p>A list of <code><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">rtracklayer::BigWigFile</a></code> objects for forward strand data (e.g., output of <code><a href="importBWlist.html">importBWlist</a></code>).</p></dd>


<dt id="arg-bwr">bwr<a class="anchor" aria-label="anchor" href="#arg-bwr"></a></dt>
<dd><p>A list of <code><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">rtracklayer::BigWigFile</a></code> objects for reverse strand data.</p></dd>


<dt id="arg-names">names<a class="anchor" aria-label="anchor" href="#arg-names"></a></dt>
<dd><p>A character vector specifying the names to assign to the resulting list of matrices.</p></dd>


<dt id="arg-grl">grl<a class="anchor" aria-label="anchor" href="#arg-grl"></a></dt>
<dd><p>A <code>GRangesList</code> where each element defines a set of genomic regions (e.g., genes, TSSs).</p></dd>


<dt id="arg-wins">wins<a class="anchor" aria-label="anchor" href="#arg-wins"></a></dt>
<dd><p>A named list specifying the number of bins (windows) for each region type in <code>grl</code>.
If multiple entries are provided, features are assumed to be combined.</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>Character string specifying the normalization mode for <code>normalizeToMatrix</code>, such as "coverage" or "mean".</p></dd>


<dt id="arg-output">output<a class="anchor" aria-label="anchor" href="#arg-output"></a></dt>
<dd><p>Output format: either <code>"norm.matrix"</code> (default) to return <code>normalizedMatrix</code> objects,
or <code>"matrix"</code> to return base R matrices.</p></dd>


<dt id="arg-strand">strand<a class="anchor" aria-label="anchor" href="#arg-strand"></a></dt>
<dd><p>Strand specificity: <code>"rev"</code> (default) for reverse alignment, <code>"for"</code> for forward, or <code>"no"</code> to ignore strand.</p></dd>


<dt id="arg-smooth">smooth<a class="anchor" aria-label="anchor" href="#arg-smooth"></a></dt>
<dd><p>Logical; if <code>TRUE</code>, applies smoothing in <code>normalizeToMatrix</code>.</p></dd>


<dt id="arg-extend">extend<a class="anchor" aria-label="anchor" href="#arg-extend"></a></dt>
<dd><p>Integer; number of base pairs to extend around each region.</p></dd>


<dt id="arg-w">w<a class="anchor" aria-label="anchor" href="#arg-w"></a></dt>
<dd><p>Optional integer; smoothing window size (used when <code>smooth = TRUE</code>).</p></dd>


<dt id="arg-include-target">include_target<a class="anchor" aria-label="anchor" href="#arg-include-target"></a></dt>
<dd><p>Logical; whether to include the target region in the matrix.</p></dd>


<dt id="arg-target-ratio">target_ratio<a class="anchor" aria-label="anchor" href="#arg-target-ratio"></a></dt>
<dd><p>Numeric between 0 and 1; controls the fraction of the matrix allocated to the target region.</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>Integer; number of bins for the matrix (used only if <code>wins</code> has length 1).</p></dd>


<dt id="arg-keep">keep<a class="anchor" aria-label="anchor" href="#arg-keep"></a></dt>
<dd><p>(Unused) Optional parameter placeholder for future development.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A named list of matrices, either as <code>EnrichedHeatmap::normalizedMatrix</code> objects
or base R numeric matrices, depending on the <code>output</code> parameter.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Load BigWig files and genomic regions</span></span></span>
<span class="r-in"><span><span class="va">forward_bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="importBWlist.html">importBWlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1.f.bw"</span>, <span class="st">"sample2.f.bw"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">reverse_bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="importBWlist.html">importBWlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1.r.bw"</span>, <span class="st">"sample2.r.bw"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">gene_grl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  gene1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">2000</span><span class="op">)</span>, strand <span class="op">=</span> <span class="st">"+"</span>, name <span class="op">=</span> <span class="st">"gene1_a"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  gene2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">3000</span>, <span class="fl">4000</span><span class="op">)</span>, strand <span class="op">=</span> <span class="st">"-"</span>, name <span class="op">=</span> <span class="st">"gene2_b"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create normalized matrix list with 20 bins</span></span></span>
<span class="r-in"><span><span class="fu">matList</span><span class="op">(</span></span></span>
<span class="r-in"><span>  bwf <span class="op">=</span> <span class="va">forward_bw</span>,</span></span>
<span class="r-in"><span>  bwr <span class="op">=</span> <span class="va">reverse_bw</span>,</span></span>
<span class="r-in"><span>  names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  grl <span class="op">=</span> <span class="va">gene_grl</span>,</span></span>
<span class="r-in"><span>  wins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Gene"</span> <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Zack Swales.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

